tests/unit/s2n_next_protocol_test.c

```diff
@@ -0,0 +1,218 @@
+
+            /* Fails for TLS1.3 */
+            server_conn->actual_protocol_version = S2N_TLS13;
+            EXPECT_FAILURE_WITH_ERRNO(s2n_next_protocol_recv(server_conn), S2N_ERR_BAD_MESSAGE);
+
+            /* Succeeds for TLS1.2 */
+            server_conn->actual_protocol_version = S2N_TLS12;
+            EXPECT_SUCCESS(s2n_next_protocol_recv(server_conn));
+            EXPECT_EQUAL(s2n_stuffer_data_available(&server_conn->handshake.io), 0);
+
+            EXPECT_BYTEARRAY_EQUAL(client_conn->application_protocol, server_conn->application_protocol, protocol_len);
+        }
+    }
+
+    /* s2n_write_npn_protocol */
+    {
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+
+        uint8_t first_protocol_len = strlen(protocols[0]);
+        EXPECT_MEMCPY_SUCCESS(client_conn->application_protocol, protocols[0], first_protocol_len + 1);
+
+        DEFER_CLEANUP(struct s2n_stuffer out = { 0 }, s2n_stuffer_free);
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+        EXPECT_SUCCESS(s2n_write_npn_protocol(client_conn, &out));
+
+        uint8_t protocol_len = 0;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &protocol_len));
+
+        uint8_t protocol[UINT8_MAX] = { 0 };
+        EXPECT_SUCCESS(s2n_stuffer_read_bytes(&out, protocol, protocol_len));
+        EXPECT_BYTEARRAY_EQUAL(protocol, protocols[0], protocol_len);
+
+        uint8_t padding_len = 0;
+        EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &padding_len));
+        EXPECT_TRUE(padding_len > 0);
+
+        for(size_t i = 0; i < padding_len; i++) {
+            uint8_t byte = UINT8_MAX;
+            EXPECT_SUCCESS(s2n_stuffer_read_uint8(&out, &byte));
+            EXPECT_EQUAL(byte, 0);
+        }
+        EXPECT_EQUAL(s2n_stuffer_data_available(&out), 0);
+    }
+
+    /* s2n_read_npn_protocol can parse s2n_write_npn_protocol */
+    {
+        DEFER_CLEANUP(struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(client_conn);
+
+        DEFER_CLEANUP(struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER), s2n_connection_ptr_free);
+        EXPECT_NOT_NULL(server_conn);
+
+        uint8_t first_protocol_len = strlen(protocols[0]);
+        EXPECT_MEMCPY_SUCCESS(client_conn->application_protocol, protocols[0], first_protocol_len + 1);
+
+        DEFER_CLEANUP(struct s2n_stuffer out = { 0 }, s2n_stuffer_free);
+        EXPECT_SUCCESS(s2n_stuffer_growable_alloc(&out, 0));
+        EXPECT_SUCCESS(s2n_write_npn_protocol(client_conn, &out));
+        EXPECT_SUCCESS(s2n_read_npn_protocol(server_conn, &out));
+
+        EXPECT_NOT_NULL(s2n_get_application_protocol(server_conn));
+        EXPECT_BYTEARRAY_EQUAL(s2n_get_application_protocol(server_conn), protocols[0], strlen(protocols[0]));
+    }
+
+    /* s2n_read_npn_protocol can read empty message */
```

Comment:
Is this allowed in the draft we're doing?

Reply:
> The "NextProtocol" message has the following format:
> 
>    struct {
>      opaque selected_protocol<0..255>;
>      opaque padding<0..255>;
>    } NextProtocol;

Yeah, I think this means the protocol can be length 0.